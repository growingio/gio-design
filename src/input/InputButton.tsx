import { CloseCircleFilled, DownFilled, EventsPresetOutlined } from '@gio-design/icons';
import { usePrefixCls } from '@gio-design/utils';
import classNames from 'classnames';
import React, { useCallback, useMemo } from 'react';
import useControlledState from '../utils/hooks/useControlledState';
import Input from './Input';
import { InputButtonProps } from './interface';

const InputButton = React.forwardRef<HTMLInputElement, InputButtonProps>((props, ref) => {
  const {
    size,
    prefixCls: customizePrefixCls,
    prefix: customizePrefix,
    suffix: customizeSuffix,
    onInputChange,
    placeholder,
    defaultValue,
    value: enterValue,
    disabled,
    hidePrefix = false,
    allowClear,
    className,
    style = {},
    maxWidth,
    ...rest
  } = props;

  const prefixCls = usePrefixCls('input-btn-new', customizePrefixCls);
  const inputCls = usePrefixCls('input-btn-new__input', customizePrefixCls);

  const [value, setValue] = useControlledState(enterValue, defaultValue);

  const onClear = useCallback(() => {
    if (disabled) {
      return;
    }
    onInputChange?.('');
    setValue('');
  }, [onInputChange, disabled, setValue]);

  const onChange = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      const inputValue = e.target.value;
      onInputChange?.(inputValue);
      setValue(inputValue);
    },
    [onInputChange, setValue]
  );

  const wrapperCls = useMemo(
    () => classNames(className, prefixCls, { [`${prefixCls}__disabled`]: disabled }),
    [className, prefixCls, disabled]
  );

  const prefix = useMemo(
    () => (hidePrefix ? null : customizePrefix || <EventsPresetOutlined />),
    [customizePrefix, hidePrefix]
  );

  const suffix = useMemo(() => {
    const hideClear = allowClear === false;
    const defaultSuffix = value && !hideClear && !disabled ? <CloseCircleFilled onClick={onClear} /> : <DownFilled />;
    return customizeSuffix || defaultSuffix;
  }, [customizeSuffix, value, onClear, allowClear, disabled]);

  const styles = maxWidth ? { maxWidth } : {};

  return (
    <span className={wrapperCls}>
      <Input
        {...rest}
        style={{ ...style, ...styles }}
        className={inputCls}
        type="button"
        value={value || placeholder}
        onChange={onChange}
        prefix={prefix}
        suffix={suffix}
        ref={ref}
        size={size}
        disabled={disabled}
      />
    </span>
  );
});

export default InputButton;
