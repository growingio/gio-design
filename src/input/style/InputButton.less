@import (reference) './index.less';

@input-btn-prefix-cls: ~'@{component-prefix}-input-btn';

.@{input-prefix-cls}.@{input-btn-prefix-cls} {
  input {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
  }
  background-color: @gray-0;
  cursor: pointer;

  &:not([disabled])::placeholder {
    color: @gray-5;
    .text-body1();
  }
  &__hover:not(&__disabled),
  &__active:not(&__disabled),
  &:not(&__disabled):hover {
    color: @blue-3;
    background-color: @gray-1;
    border: 1px solid @gray-2;

    .@{input-prefix-cls}__prefix {
      svg {
        color: @blue-3;
        // cursor: not-allowed;
      }
    }
    input {
      color: @blue-3;
      cursor: pointer;
    }
  }
  &:is([disabled]):hover {
    color: @gray-3;
    border: 1px solid @gray-2;
  }

  &:not(&__disabled):focus {
    border: 1px solid @gray-2;
    outline: none;
  }
  &__disabled {
    background-color: @gray-0;
    cursor: not-allowed;
    input {
      color: @gray-3;
    }
    svg {
      color: @gray-3;
      cursor: not-allowed;
    }

    &:hover {
      .@{input-btn-prefix-cls} {
        color: @gray-3;
        background-color: @gray-0;
        cursor: not-allowed;
      }
      .@{input-prefix-cls} {
        &__prefix {
          color: @gray-3;
          cursor: not-allowed;
          svg {
            color: @gray-3;
            cursor: not-allowed;
          }
        }

        &__suffix {
          cursor: not-allowed;
        }
      }
    }

    .@{input-btn-prefix-cls} {
      background-color: @gray-0;
    }
    .@{input-prefix-cls} {
      &__prefix-disabled {
        color: @gray-3;
        cursor: not-allowed;
        svg {
          color: @gray-3;
          cursor: not-allowed;
        }
      }
    }
  }
}
