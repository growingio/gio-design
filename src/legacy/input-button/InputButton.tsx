import { CloseCircleFilled, EventsPresetOutlined, SearchOutlined } from '@gio-design/icons';
import { usePrefixCls } from '@gio-design/utils';
import classNames from 'classnames';
import React, { useCallback, useMemo, useState } from 'react';
import Input from '../input/Input';
import { InputButtonProps } from './interface';

import './style';

const InputButton = React.forwardRef<HTMLInputElement, InputButtonProps>((props, ref) => {
  const {
    prefixCls: customizePrefixCls,
    prefix,
    onChange: onChangeFC,
    value: enterValue,
    disabled,
    placeholder,
  } = props;

  const prefixCls = usePrefixCls('search', customizePrefixCls);
  const [value, setValue] = useState(enterValue);

  const [canClear, setClear] = useState(!!enterValue);

  const prefixIcon = <EventsPresetOutlined />;

  const suffixCls = useMemo(
    () =>
      classNames(`${prefixCls}__suffix`, {
        [`${prefixCls}__suffix-clear`]: canClear,
        [`${prefixCls}__suffix-disabled`]: disabled,
      }),
    [prefixCls, canClear, disabled]
  );

  const suffix = useMemo(
    () => (canClear ? <CloseCircleFilled className={suffixCls} /> : <SearchOutlined className={suffixCls} />),
    [suffixCls, canClear]
  );

  const onChange = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
    const inputValue = e.target.value;
    onChangeFC && onChangeFC(e);
    // onSearch && onSearch(inputValue);
    setValue(inputValue);
    setClear(!!inputValue);
  }, []);

  const resetPlaceholder = useMemo(() => (placeholder || disabled ? '无法搜索' : '搜索'), [placeholder, disabled]);
  return (
    <Input
      {...props}
      placeholder={resetPlaceholder}
      value={value}
      onChange={onChange}
      prefix={prefixIcon}
      suffix={suffix}
      ref={ref}
    />
  );
});

export default InputButton;
