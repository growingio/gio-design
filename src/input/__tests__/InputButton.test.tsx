import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import InputButton from '../InputButton';

describe('testing InputButton', () => {
  it('should render an element input[type="button"]', () => {
    const { container } = render(<InputButton defaultValue='111' />);
    expect(container.querySelector('input[type="button"]')).toBeTruthy()
  });
  it('allowClear', () => {
    let changedValue;
    const mockChange = jest.fn(e => {
      changedValue = e;
    });
    const { container } = render(<InputButton defaultValue='111' allowClear onChange={mockChange} />);

    expect(container.querySelector('span[aria-label="error-filled"]')).toBeTruthy();
    fireEvent.click(container.querySelector('span[aria-label="error-filled"]'));
    expect(mockChange).toHaveBeenCalled();
    expect(changedValue).toEqual('');
  });
  it('should has active class when set active=true ', () => {
    const { container } = render(<InputButton defaultValue='111' active />);
    expect(container.querySelector('.gio-input-btn__active')).toBeTruthy();
  })
  it('diabled should not allowClear', () => {
    const { container } = render(<InputButton defaultValue='111' allowClear disabled />);
    expect(container.querySelector('.gio-input__suffix-disabled')).toBeTruthy();
    expect(container.querySelector('span[aria-label="error-filled"]')).toBeFalsy();
    expect(container.querySelector('span[aria-label="down-filled"]')).toBeTruthy();

  });
  it('trigger event when click clear icon', () => {
    const mockClear = jest.fn();
    const mockInputChange = jest.fn()
    const { container } = render(<InputButton defaultValue='111' allowClear onClear={mockClear} onInputChange={mockInputChange} />);
    fireEvent.click(container.querySelector('span[aria-label="error-filled"]'));
    expect(mockClear).toHaveBeenCalled();
    expect(mockInputChange).toHaveBeenCalledWith('');
  });
  it('custom style and maxWidth', () => {
    const { container } = render(<InputButton defaultValue='111' style={{ color: 'red' }} allowClear maxWidth={200} />);
    expect(container.querySelector('.gio-input-btn__input')).toHaveStyle({ color: 'red', maxWidth: '200px' });
  })
})