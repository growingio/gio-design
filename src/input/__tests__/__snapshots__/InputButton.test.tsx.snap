// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing InputButton custom style and maxWidth 1`] = `
<div>
  <span
    class="gio-input-btn__input gio-input-btn gio-input__hover gio-input"
    style="color: red; max-width: 200px;"
  >
    <input
      data-testid="input-button"
      title="111"
      type="button"
      value="111"
    />
    <div
      class="gio-input__suffix"
    >
      <span
        aria-label="error-filled"
        class="gio-icon"
        role="img"
      >
        <svg
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="0 0 14 14"
          width="1em"
        >
          <path
            clip-rule="evenodd"
            d="M7 0c3.863 0 7 3.137 7 7s-3.137 7-7 7-7-3.137-7-7 3.137-7 7-7ZM4.172 4.17a.499.499 0 0 1 .706 0l1.98 1.982a.2.2 0 0 0 .283 0L9.122 4.17a.499.499 0 1 1 .707.707l-1.98 1.98a.2.2 0 0 0 0 .283l1.98 1.981a.499.499 0 1 1-.707.707l-1.98-1.98a.2.2 0 0 0-.283 0l-1.981 1.98a.499.499 0 1 1-.707-.707l1.98-1.98a.2.2 0 0 0 0-.283l-1.98-1.981a.499.499 0 0 1 .001-.707Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </div>
  </span>
</div>
`;
